crates/matrix-sdk/src/client/futures.rs

```diff
@@ -69,26 +80,61 @@ where
+            if let Err(Some(ErrorKind::UnknownToken { soft_logout })) =
+                res.as_ref().map_err(HttpError::client_api_error_kind)
+            {
+                trace!("Token refresh: Unknown token error received.");
+                // If automatic token refresh isn't supported, there is nothing more to do.
+                if !client.inner.handle_refresh_tokens {
+                    trace!("Token refresh: Automatic refresh disabled.");
+                    client.broadcast_unknown_token(soft_logout);
+                    return res;
+                }
+
+                // Try to refresh the token and retry the request.
+                if let Err(refresh_error) = client.refresh_access_token().await {
+                    match &refresh_error {
+                        RefreshTokenError::RefreshTokenRequired => {
+                            trace!("Token refresh: The session doesn't have a refresh token.");
+                            // Refreshing access tokens is not supported by this `Session`, ignore.
+                            client.broadcast_unknown_token(soft_logout);
+                        }
+                        RefreshTokenError::Oidc(oidc_error) => {
+                            let oidc_error = oidc_error.deref();
+                            match oidc_error {
+                                OidcError::Oidc(OidcClientError::TokenRefresh(
+                                    TokenRefreshError::Token(TokenRequestError::Http(
+                                        OidcHttpError {
+                                            body:
+                                                Some(OidcErrorBody {
+                                                    error: ClientErrorCode::InvalidGrant,
+                                                    ..
+                                                }),
+                                            ..
+                                        },
+                                    )),
+                                )) => {
+                                    trace!("Token refresh: OIDC refresh denied.");
+                                    // The refresh was denied, signal to sign out the user.
+                                    client.broadcast_unknown_token(soft_logout);
+                                }
+                                _ => {
+                                    trace!("Token refresh: OIDC refresh encountered a problem.");
+                                    // The refresh failed for other reasons, no
+                                    // need to sign out.
+                                }
+                            };
+                            return Err(refresh_error.into());
+                        }
+                        _ => {
+                            trace!("Token refresh: Token refresh failed.");
+                            return Err(refresh_error.into());
```

Comment:
I presume there's also an error type for homeserver refresh that matches `InvalidGrant` to signal that the refresh was specifically denied. If so this should be handled here.

Reply:
According to the spec:

> If the token refresh fails and the error response included a soft_logout: true property, then the client can treat it as a [soft logout](https://spec.matrix.org/unstable/client-server-api/#soft-logout) and attempt to obtain a new access token by re-logging in. If the error response does not include a soft_logout: true property, the client should consider the user as being logged out.

So the error doesn't matter, it's always considered a logout. I'm wondering if we should do the same for OIDC, because if the refresh token fails endlessly no request will be able to be made.

Reply:
I'm going to defer this one to @hughns who suggested we should only use `InvalidGrant` as the signal to logout the user for OIDC.

Reply:
It's quite a while since I last discussed soft logout with @sandhose so my memory is hazy on it...


But, I suspect it is the case that Matrix's soft logout concept would no longer be possible.

Instead -

If a client founds itself with an invalid refresh token - most likely meaning that the session was signed out from another client or My Account UI - then it could ask the user whether they want to attempt re-authenticate or whether to Logout completely.

If the user asks to re-authenticate then you can reuse the existing device ID in the `urn:matrix:client:device:XXX` scope in the authorization request.

If the user says Logout completely then you would destroy all the local data.

Reply:
Ok, so for now, using `InvalidGrant` is correct for OIDC. Given that native refresh is going to sign the user out on any error in the current implementation, I'm just going to update this PR to do the same with a note that it isn't correct.